src/components/CommandBar/CommandBar.tsx

```diff
@@ -97,26 +97,33 @@ export class CommandBar extends React.Component<ICommandBarProps, ICommandBarSta
       );
     }
 
+    let visibleOverflow = false;
+    if (renderedOverflowItems) {
+      visibleOverflow = renderedOverflowItems.some(function(item) {
+        return (item.icon !== '' || item.name !== '');
+      });
+    }
+
     return (
       <div className={ css('ms-CommandBar', className) } ref='commandBarRegion'>
         { searchBox }
         <FocusZone ref='focusZone' direction={ FocusZoneDirection.horizontal } rootProps={ { role: 'menubar' } }>
           <div className='ms-CommandBar-primaryCommands' ref='commandSurface'>
             { renderedItems.map((item, index) => (
               this._renderItemInCommandBar(item, index, expandedMenuItemKey)
-            )).concat((renderedOverflowItems && renderedOverflowItems.length) ? [
-            <div className='ms-CommandBarItem' key={ OVERFLOW_KEY } ref={ OVERFLOW_KEY }>
-              <button
-                id={ this._id + OVERFLOW_KEY }
-                className={ css('ms-CommandBarItem-link', { 'is-expanded': (expandedMenuItemKey === OVERFLOW_KEY) }) }
-                onClick={ this._onOverflowClick }
-                role='menuitem'
-                aria-label={ this.props.elipisisAriaLabel || '' }
-                aria-haspopup={ true }
-              >
-                <i className='ms-CommandBarItem-overflow ms-Icon ms-Icon--More' />
-              </button>
-            </div>
+            )).concat((renderedOverflowItems && renderedOverflowItems.length && visibleOverflow) ? [
```

Comment:
If a pure CSS solution to this bug is possible that will be better, wherein an item does not take up space on the bar if it doesn't contain both icon and name. I am against putting a hard coded restriction..